### ISSUE TITLE
Unable to Upload Files to Bucket using JQuery - CROS blocking

### ISSUE THREAD

**Issue by RageshAntony** [Backblaze/B2_Command_Line_Tool#536]
Hi ,

We tried to upload files to our backblaze b2 bucket , but we get blocked by CORS rules .

The problem is we enabled CORS in the CORS Rules settings in the bucket setting as .....

    Share everything in this bucket with every origin.

But still we getting CORS blocked the request . We tried many times ....

The Browser Console states as this ::

    Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://pod-000-1111-11.backblaze.com/b2api/v2/b2_upload_file/e6eb663b7f05aaa86a790e15/c002_v0001111_t0055. (Reason: CORS preflight channel did not succeed).
    Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://pod-000-1111-11.backblaze.com/b2api/v2/b2_upload_file/e6eb663b7f05aaa86a790e15/c002_v0001111_t0055. (Reason: CORS request did not succeed                    

Here is our code ::

```

    <html>
        <body>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
    <link href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css" rel="stylesheet"/>
    <link href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/js/bootstrap.min.js" rel="stylesheet" type="text/css" > 


    <form method="POST" action="">
    	
    	<button class="btn btn-lg upload-btn" id="upload-btn" type="button">Upload File</button>
    	<input type="file" class="form-control" name="data" id="upload-input" ><br/><br/><br/>
    	<div class="progress">
    		<div class="progress-bar" role="progressbar">0%</div>
    	</div> 
    </form>


    <script>
    $(document).ready(function() {

    	$('#upload-input').change(function(){
    		$('.progress-bar').text('0%');
    		$('.progress-bar').width('0%');
    	});
    	$('#upload-btn').on('click', function() {
    		$('.progress-bar').text('0%');
    		$('.progress-bar').width('0%');
    		var uploadInput = $('#upload-input'); 

    		if (uploadInput[0].files[0] != undefined) {
    			var formData = new FormData();
    			formData.append('data', uploadInput[0].files[0]);

    			$.ajax({
    				url: 'https://pod-000-1111-11.backblaze.com/b2api/v2/b2_upload_file/e6eb663b7f05aaa86a790e15/c002_v0001111_t0055',
                    type: 'POST',
                    crossDomain: true,
                    data: formData,
                    headers: {
        "Authorization": "4_0026b6bf5a8a9e50000000001_0189298a_a56c72_upld_RRvm_eYswwXtCr5mva4BZnGEBW0=",
        "X-Bz-File-Name": "Client Proposed Structure .pdf",
       "Access-Control-Allow-Origin": "*",
       "Access-Control-Allow-Headers": "Origin, X-Requested-With, Content-Type, Accept,X-Bz-File-Name,X-Bz-Content-Sha1,X-Bz-File-Name,Authorization",
       "Access-Control-Request-Method" : "POST",
        "Content-Type": "application/pdf",
        "X-Bz-Content-Sha1": "do_not_verify",

      },
    				processData: false,
    				contentType: false,

    				success: function(data) {
    					$('#upload-input').val('');
    				},
    				xhr: function() {
    					var xhr = new XMLHttpRequest();
    					xhr.upload.addEventListener("progress", function(e) {
    						if (e.lengthComputable) {
    							//var uploadPercent = e.loaded / e.total; typo uploadpercent (all lowercase)
    							var uploadpercent = e.loaded / e.total; 
    							uploadpercent = (uploadpercent * 100); //optional Math.round(uploadpercent * 100)
    							$('.progress-bar').text(uploadpercent + '%');
    							$('.progress-bar').width(uploadpercent + '%');
    							if (uploadpercent == 100) {
    								$('.progress-bar').text('Completed');
    							}
    						}
    					}, false);
    					
    					return xhr;
    				}})}});
    });
    </script></body></html>
```

Please help us to resolve this issue

**Comment by ppolewicz** [comment]
This is not a b2 CLI issue, but since you are already here, perhaps I can give you a hint.

Try to connect your application (or browser, not sure what you are doing) through mitmproxy or use Chrome F12 network grabber or something like that, to intercept the CORS request and see why it failed.

**Comment by RageshAntony** [comment]
Ok I will try . Thanks for the response 

/// This is not a b2 CLI issue, ///
Yeah . I know . But I can't find suitable category to ask this question . Thats why I asked here .

Where I have to ask question regarding BackBlaze B2 Web APIs ?

**Comment by ppolewicz** [comment]
normally you should ask b2 support [here](https://help.backblaze.com/hc/en-us/requests/new)

**Comment by RageshAntony** [comment]
> normally you should ask b2 support [here](https://help.backblaze.com/hc/en-us/requests/new)

I already asked there . They told to include auth token in url not in header . I done that . But still issue persists 

So I replied to the email and waiting still now

Since the time is short , I asked it here .

**Comment by RageshAntony** [comment]
![image](https://user-images.githubusercontent.com/36100215/50515674-1a28d300-0acc-11e9-8e33-a3baa3ffe2d9.png)

![image](https://user-images.githubusercontent.com/36100215/50515688-2ad94900-0acc-11e9-9fbd-30aaf1170757.png)

These are the errors in console

**Comment by certainmagic** [comment]
Hi —

The choice that says “Share everything in this bucket with every origin.”
is for read-only sharing. To enable cross-origin uploads, you need to set
custom CORS rules.

To upload, you need to allow the appropriate upload operations by adding
them to the rule’s ‘allowedOperations’.

You can learn more about CORS here, especially in the “CORS for uploads
section”:

https://www.backblaze.com/b2/docs/cors_rules.html

You can use the b2 CLI to see and change your CORS rules. Use “b2
list-buckets” to see your rules and “b2 update-bucket —corsRules ...” to
change them.  The syntax for setting the rules with “b2 update-bucket” can
be a little tricky since you need include the JSON for the rules on the
command line, so you have to be careful with the quoting. There are bash
techniques for inlining a file, but I’m not at a computer, so I won’t try
to suggest one now. Maybe someone else can chime in.

ttfn,
ab

**Comment by ppolewicz** [comment]
It doesn't look like you have included the auth token in the url. I think it may be the reason preflight (`OPTIONS`) request is failing.

**Comment by RageshAntony** [comment]
I already included it like this 

formdata.append('Authorization','tokenid');

But not worked

Will you please tell the correct format because i also tried the formats 

 - endpoint?Authorization=tokenid
 - endpoint/Authorization/tokenid

Nothing worked

**Comment by RageshAntony** [comment]
@certainmagic 

You told a new solution. I will try and update here

**Comment by RageshAntony** [comment]
> It doesn't look like you have included the auth token in the url. I think it may be the reason preflight (`OPTIONS`) request is failing.

 @ppolewicz 

How to include Auth token in the preflight OPTIONS request in jQuery Ajax request ?

**Comment by RageshAntony** [comment]
As suugested by @certainmagic I succesfuly bypassed cors 

So now no CORS failed error

But now get a new error
![image](https://user-images.githubusercontent.com/36100215/50535144-cae6af00-0b6c-11e9-98a0-e5d2277ed2c9.png)
=====
![image](https://user-images.githubusercontent.com/36100215/50535324-0aae9600-0b6f-11e9-859b-98c9f98dbded.png)


I got this above error ::

below is my jquery request 

		```
$('#upload-btn').on('click', function() {

			var uploadInput = $('#upload-input'); 

	var form = new FormData();
form.append("data", uploadInput[0].files[0]);
/*
var settings = {
  "async": true,
  "crossDomain": true,
  "url": "https://pod-000-1112-10.backblaze.com/b2api/v2/b2_upload_file/e6eb663b7f05aaa86a790e15/c002_v0001112_t0041",
  "method": "POST",
  "headers": {
    "Authorization": "4_0026b6bf5a8a9e50000000001_01893506_cc0e42_upld_mWKyh7WuNdag6q4mHDJ6m-U4yHk=",
    "X-Bz-File-Name": "Client Proposed Structure .pdf",

    "Content-Type": "application/pdf",
    "X-Bz-Content-Sha1": "do_not_verify",
  },
  "processData": false,
  "contentType": false,
  "mimeType": "multipart/form-data",
  "data": form
}

$.ajax(settings).done(function (response) {
  console.log(response);
  alert(response);
});

});


```

**Comment by RageshAntony** [comment]
Hi guys 

Finally I succeeded in uploading files from browser to backblaze storage

But now a weired bug occurs 

The uploaded files are listing in  CLI (b2 list_files ) But no in the Backblaze web site but files count showing correct

here the images 
**The file listed in CLI**
![image](https://user-images.githubusercontent.com/36100215/50536974-751c0200-0b80-11e9-8efa-bb0400a9e51e.png)

But no found in Web Console (see the text search in top right corner ). I refreshed many times 

![image](https://user-images.githubusercontent.com/36100215/50537004-b6acad00-0b80-11e9-806c-88597ede8298.png)

But the file count is correct . But when I enter in to the bucket , it showing less files (count - 1)

![image](https://user-images.githubusercontent.com/36100215/50537047-228f1580-0b81-11e9-8ffa-7438c1d0e65d.png)

This problem resolves only after signing out and signing in again .

Please look after this issue

**Comment by ppolewicz** [comment]
Again, this is not a B2 CLI issue. It looks like a problem with cache invalidation on the b2 web admin site, please report it to b2 support (as a separate ticket).

I'm happy to see that you have successfully achieved your goal.

Please close the ticket when this discussion is concluded.

**Comment by RageshAntony** [comment]
please report it to b2 support (as a separate ticket). /// Yes . Already done

**Comment by RageshAntony** [comment]
Thanks to @ppolewicz and @certainmagic for helping me

And special thanks to @certainmagic since he told about Custom CORS . Then only CORS issue solved 

Bye for now
